File: src/FluentCards/InputChoiceSetBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating InputChoiceSet elements.
/// </summary>
public class InputChoiceSetBuilder
{
    private readonly InputChoiceSet _input = new() { Choices = new List<Choice>() };

    /// <summary>
    /// Sets the unique identifier for the input.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithId(string id)
    {
        _input.Id = id;
        return this;
    }

    /// <summary>
    /// Sets the label for the input.
    /// </summary>
    /// <param name="label">The input label.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithLabel(string label)
    {
        _input.Label = label;
        return this;
    }

    /// <summary>
    /// Sets the style for displaying choices.
    /// </summary>
    /// <param name="style">The choice input style.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithStyle(ChoiceInputStyle style)
    {
        _input.Style = style;
        return this;
    }

    /// <summary>
    /// Sets whether multiple choices can be selected.
    /// </summary>
    /// <param name="multiSelect">True for multi-select, false for single select.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder IsMultiSelect(bool multiSelect = true)
    {
        _input.IsMultiSelect = multiSelect;
        return this;
    }

    /// <summary>
    /// Sets the default selected value(s).
    /// </summary>
    /// <param name="value">The default value.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithValue(string value)
    {
        _input.Value = value;
        return this;
    }

    /// <summary>
    /// Sets the placeholder text for the input.
    /// </summary>
    /// <param name="placeholder">The placeholder text.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithPlaceholder(string placeholder)
    {
        _input.Placeholder = placeholder;
        return this;
    }

    /// <summary>
    /// Sets whether the input is required.
    /// </summary>
    /// <param name="required">True if required, false otherwise.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder IsRequired(bool required = true)
    {
        _input.IsRequired = required;
        return this;
    }

    /// <summary>
    /// Sets the error message to display when validation fails.
    /// </summary>
    /// <param name="message">The error message.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithErrorMessage(string message)
    {
        _input.ErrorMessage = message;
        return this;
    }

    /// <summary>
    /// Sets whether choice text should wrap.
    /// </summary>
    /// <param name="wrap">True to wrap, false otherwise.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithWrap(bool wrap)
    {
        _input.Wrap = wrap;
        return this;
    }

    /// <summary>
    /// Adds a choice to the input.
    /// </summary>
    /// <param name="title">The display text for the choice.</param>
    /// <param name="value">The internal value for the choice.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder AddChoice(string title, string value)
    {
        _input.Choices!.Add(new Choice { Title = title, Value = value });
        return this;
    }

    /// <summary>
    /// Adds a choice to the input.
    /// </summary>
    /// <param name="choice">The choice to add.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder AddChoice(Choice choice)
    {
        _input.Choices!.Add(choice);
        return this;
    }

    /// <summary>
    /// Sets the position of the label relative to the input.
    /// </summary>
    /// <param name="position">The label position.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithLabelPosition(InputLabelPosition position)
    {
        _input.LabelPosition = position;
        return this;
    }

    /// <summary>
    /// Sets the width of the label when displayed inline.
    /// </summary>
    /// <param name="width">The label width (e.g., "40%" or a pixel value).</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithLabelWidth(string width)
    {
        _input.LabelWidth = width;
        return this;
    }

    /// <summary>
    /// Sets the visual style of the input.
    /// </summary>
    /// <param name="style">The input style.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public InputChoiceSetBuilder WithInputStyle(InputStyle style)
    {
        _input.InputStyle = style;
        return this;
    }

    /// <summary>
    /// Builds and returns the configured InputChoiceSet.
    /// </summary>
    /// <returns>The configured InputChoiceSet instance.</returns>
    public InputChoiceSet Build()
    {
        return _input;
    }
}

File: go/fluentcards/input_choice_set_builder.go
package fluentcards

// InputChoiceSetBuilder builds an Input.ChoiceSet Adaptive Card element.
type InputChoiceSetBuilder struct {
	data map[string]any
}

func newInputChoiceSetBuilder() *InputChoiceSetBuilder {
	return &InputChoiceSetBuilder{data: map[string]any{"type": "Input.ChoiceSet", "id": "", "choices": []any{}}}
}

func (b *InputChoiceSetBuilder) WithID(id string) *InputChoiceSetBuilder {
	b.data["id"] = id
	return b
}

func (b *InputChoiceSetBuilder) WithLabel(label string) *InputChoiceSetBuilder {
	b.data["label"] = label
	return b
}

func (b *InputChoiceSetBuilder) WithPlaceholder(placeholder string) *InputChoiceSetBuilder {
	b.data["placeholder"] = placeholder
	return b
}

func (b *InputChoiceSetBuilder) WithValue(value string) *InputChoiceSetBuilder {
	b.data["value"] = value
	return b
}

func (b *InputChoiceSetBuilder) WithStyle(style ChoiceInputStyle) *InputChoiceSetBuilder {
	b.data["style"] = string(style)
	return b
}

func (b *InputChoiceSetBuilder) WithIsMultiSelect(isMultiSelect bool) *InputChoiceSetBuilder {
	b.data["isMultiSelect"] = isMultiSelect
	return b
}

func (b *InputChoiceSetBuilder) WithWrap(wrap bool) *InputChoiceSetBuilder {
	b.data["wrap"] = wrap
	return b
}

func (b *InputChoiceSetBuilder) WithIsRequired(isRequired bool) *InputChoiceSetBuilder {
	b.data["isRequired"] = isRequired
	return b
}

func (b *InputChoiceSetBuilder) WithErrorMessage(errorMessage string) *InputChoiceSetBuilder {
	b.data["errorMessage"] = errorMessage
	return b
}

func (b *InputChoiceSetBuilder) WithSpacing(spacing Spacing) *InputChoiceSetBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

func (b *InputChoiceSetBuilder) WithIsVisible(isVisible bool) *InputChoiceSetBuilder {
	b.data["isVisible"] = isVisible
	return b
}

func (b *InputChoiceSetBuilder) WithSeparator(separator bool) *InputChoiceSetBuilder {
	b.data["separator"] = separator
	return b
}

func (b *InputChoiceSetBuilder) WithHeight(height string) *InputChoiceSetBuilder {
	b.data["height"] = height
	return b
}

func (b *InputChoiceSetBuilder) WithFallback(fallback any) *InputChoiceSetBuilder {
	b.data["fallback"] = fallback
	return b
}

func (b *InputChoiceSetBuilder) WithRequires(key, version string) *InputChoiceSetBuilder {
	reqs, ok := b.data["requires"].(map[string]any)
	if !ok {
		reqs = map[string]any{}
	}
	reqs[key] = version
	b.data["requires"] = reqs
	return b
}

func (b *InputChoiceSetBuilder) WithRtl(rtl bool) *InputChoiceSetBuilder {
	b.data["rtl"] = rtl
	return b
}

func (b *InputChoiceSetBuilder) WithLabelPosition(position InputLabelPosition) *InputChoiceSetBuilder {
	b.data["labelPosition"] = string(position)
	return b
}

func (b *InputChoiceSetBuilder) WithLabelWidth(width string) *InputChoiceSetBuilder {
	b.data["labelWidth"] = width
	return b
}

func (b *InputChoiceSetBuilder) WithInputStyle(style InputStyle) *InputChoiceSetBuilder {
	b.data["inputStyle"] = string(style)
	return b
}

// AddChoiceadds a choice with the given title and value strings.
func (b *InputChoiceSetBuilder) AddChoice(title, value string) *InputChoiceSetBuilder {
	choices := b.data["choices"].([]any)
	b.data["choices"] = append(choices, map[string]any{"title": title, "value": value})
	return b
}

// AddChoiceMap adds a pre-built choice map directly.
func (b *InputChoiceSetBuilder) AddChoiceMap(choice map[string]any) *InputChoiceSetBuilder {
	choices := b.data["choices"].([]any)
	b.data["choices"] = append(choices, choice)
	return b
}

// WithChoicesData sets a dynamic data query for fetching choices from a data source (Adaptive Cards 1.6+).
// dataset is the dataset identifier, e.g. "graph.microsoft.com/users".
func (b *InputChoiceSetBuilder) WithChoicesData(dataset string) *InputChoiceSetBuilder {
	b.data["choices.data"] = map[string]any{"type": "Data.Query", "dataset": dataset}
	return b
}

// WithChoicesDataFull sets a dynamic data query with dataset, count, and skip parameters.
func (b *InputChoiceSetBuilder) WithChoicesDataFull(dataset string, count int, skip int) *InputChoiceSetBuilder {
	b.data["choices.data"] = map[string]any{"type": "Data.Query", "dataset": dataset, "count": count, "skip": skip}
	return b
}

func (b *InputChoiceSetBuilder) Build() Card {
	return b.data
}
